### PR TITLE
Fix ownership of CFSocketCreateRunLoopSource

### PR DESCRIPTION
**Pull request by Therzok**

### CODE REVIEW

**Comment by Therzok** [comment]
I just noticed something. `CFSocketInvalidate` is not called, so the runloopsource is never removed. So the native leak of CFSocket will still be there. Should we bind CFSocket Invalidate to allow users to invalidate the socket before disposal?

**Comment by Therzok** [comment]
There's a few other cans of worms I won't open in this PR:
a) The info pointer should actually contain a release call to the GCHandle, not Dispose
b) The info pointer is actually the wrong type.

File: src/CoreFoundation/CFSocket.cs [hunk 1]
```diff
@@ -209,9 +212,62 @@ static byte[] CreateData (IPEndPoint endpoint)
...
+	}
+
 	public class CFSocket : CFType, INativeObject, IDisposable {
 		IntPtr handle;
 		GCHandle gch;
+		int retainCount = 1;
+
+		internal void Retain ()
+		{
+			++retainCount;
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFSocket.cs]: This should use `Interlocked.Increment`

File: src/CoreFoundation/CFSocket.cs [hunk 2]
```diff
@@ -209,9 +212,62 @@ static byte[] CreateData (IPEndPoint endpoint)
...
+		}
+
+		internal void Release ()
+		{
+			if (--retainCount == 0 && gch.IsAllocated) {
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFSocket.cs]: And `Interlocked.Decrement`.

File: src/CoreFoundation/CFSocket.cs [hunk 3]
```diff
@@ -300,90 +355,68 @@ public CFSocket (AddressFamily family, SocketType type, ProtocolType proto)
...
-				Marshal.FreeHGlobal (ptr);
-			}
-
-			if (handle == IntPtr.Zero)
-				throw new CFSocketException (CFSocketError.Error);
-			gch = GCHandle.Alloc (this);
-
-			var source = new CFRunLoopSource (CFSocketCreateRunLoopSource (IntPtr.Zero, handle, 0));
-			loop.AddSource (source, CFRunLoop.ModeDefault);
+			var ctx = new CFSocketContext (this);
```
Comment (rolfbjarne) [inline src/CoreFoundation/CFSocket.cs]: You can move the creation of the `CFSocketContext` into the `Initialize` method as well, it's identical in all three cases.

File: tests/monotouch-test/CoreFoundation/SocketTest.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+namespace MonoTouchFixtures.CoreFoundation {
+
+	[TestFixture]
+	[Preserve (AllMembers = true)]
+	public class CFSocketTest {
+		static readonly byte[] dataToSend = { 0, 1, 2, 3 };
+
+		[Test]
+		public void RetainCount ()
```
Comment (rolfbjarne) [inline tests/monotouch-test/CoreFoundation/SocketTest.cs]: I think it's also possible to create a test that checks if the CFSocket instance is actually collected by the GC:

1. Create several CFSockets on a background thread, create a weak gchandle to each, and then dispose them.
2. On the main thread, wait for the background thread to finish.
3. Run the GC a few times.
4. Assert that at least one of the weak gchandles doesn't have a target anymore.

This test should ensure that the refcounting logic for the context is correct.

File: src/CoreFoundation/CFSocket.cs (hunk 3 above)
Comment (Therzok) [reply src/CoreFoundation/CFSocket.cs]: How? We depend on the handle form the PInvoke. And the pinvoke depends on the struct.

File: tests/monotouch-test/CoreFoundation/SocketTest.cs (hunk 4 above)
Comment (Therzok) [reply tests/monotouch-test/CoreFoundation/SocketTest.cs]: Did manual tracing on the unit test, it was collecting them. But sure, can add that.

File: src/CoreFoundation/CFSocket.cs (hunk 3 above)
Comment (rolfbjarne) [reply src/CoreFoundation/CFSocket.cs]: Ah duh, nvm.

**Comment by Therzok** [comment]
Thanks for taking over :heart:

**Review by rolfbjarne — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]
❤️

File: src/CoreFoundation/CFSocket.cs [hunk 5]
```diff
@@ -272,28 +350,31 @@ static void OnCallback (IntPtr s, nuint type, IntPtr address, IntPtr data, IntPt
...
+				using (var cfdata = new CFData (data, false)) {
+					if (cfdata.Length > 0) {
+						var ep = CFSocketAddress.EndPointFromAddressPtr (address, false)!;
+						socket.OnData (new CFSocketDataEventArgs (ep, cfdata.GetBuffer ()));
+					}
+				}
 			} else if (cbType == CFSocketCallBackType.NoCallBack) {
 				// nothing to do
-			} else if (cbType == CFSocketCallBackType.ReadCallBack) {
+			} else if (cbType == CFSocketCallBackType.ReadCallBack && socket.ReadEvent != null) {
```
Comment (mandel-macaque) [inline src/CoreFoundation/CFSocket.cs:362]:
```suggestion
			} else if (cbType == CFSocketCallBackType.ReadCallBack && socket.ReadEvent is not null) {
```

File: src/CoreFoundation/CFSocket.cs [hunk 6]
```diff
@@ -272,28 +350,31 @@ static void OnCallback (IntPtr s, nuint type, IntPtr address, IntPtr data, IntPt
...
 				socket.OnRead (new CFSocketReadEventArgs ());
-			} else if (cbType == CFSocketCallBackType.WriteCallBack) {
+			} else if (cbType == CFSocketCallBackType.WriteCallBack && socket.WriteEvent != null) {
```
Comment (mandel-macaque) [inline src/CoreFoundation/CFSocket.cs:364]:
```suggestion
			} else if (cbType == CFSocketCallBackType.WriteCallBack && socket.WriteEvent is not null) {
```

**Review by mandel-macaque — APPROVED** [review]

**Pull request merged**